src/compas/geometry/_primitives/torus.py

```diff
@@ -0,0 +1,281 @@
+from __future__ import print_function
+from __future__ import absolute_import
+from __future__ import division
+
+from compas.geometry import Plane
+
+__all__ = ['Torus']
+
+
+class Torus(object):
+    """A torus is defined by a plane and two radii.
+
+    Attributes
+    ----------
+    plane : :class:`compas.geometry.Plane`
+        The plane of the torus.
+    radius1: float
+        The radius of the axis.
+    radius2: float
+        The radius of the pipe.
+
+    Examples
+    --------
+    >>> from compas.geometry import Plane
+    >>> from compas.geometry import Torus
+    >>> torus = Torus(Plane.worldXY(), 5., 2.)
+
+    """
+
+    __slots__ = ['_plane', '_radius1', '_radius2']
+
+    def __init__(self, plane, radius1, radius2):
+        self._plane = None
+        self._radius1 = None
+        self._radius2 = None
+        self.plane = plane
+        self.radius1 = radius1
+        self.radius2 = radius2
+
+    @classmethod
+    def from_data(cls, data):
+        """Construct a torus from its data representation.
+
+        Parameters
+        ----------
+        data : :obj:`dict`
+            The data dictionary.
+
+        Returns
+        -------
+        Torus
+            The constructed torus.
+
+        Examples
+        --------
+        >>> from compas.geometry import Torus
+        >>> data = {'plane': Plane.worldXY().data, 'radius1': 4., 'radius2': 1.}
+        >>> torus = Torus.from_data(data)
+
+        """
+        torus = cls(Plane.worldXY(), 1, 1)
+        torus.data = data
+        return torus
+
+    # ==========================================================================
+    # descriptors
+    # ==========================================================================
+
+    @property
+    def plane(self):
+        """Plane: The torus' plane."""
+        return self._plane
+
+    @plane.setter
+    def plane(self, plane):
+        self._plane = Plane(plane[0], plane[1])
+
+    @property
+    def radius1(self):
+        """float: The radius of the axis."""
+        return self._radius1
+
+    @radius1.setter
+    def radius1(self, radius):
+        self._radius1 = float(radius)
+    
+    @property
+    def radius2(self):
+        """float: The radius of the pipe."""
+        return self._radius2
+
+    @radius2.setter
+    def radius2(self, radius):
+        self._radius2 = float(radius)
+
```

Comment:
would add `inner_radius` and `outer_radius` to the properties and maybe also add `Torus.from_inner_and_outer_radius()`,  as it seems to be often used as constuctor.